### PR TITLE
Add cache for entity credentials

### PR DESCRIPTION
**Pull request by ThumulaPerera**
### Purpose

This pull request enhances the caching layer for the entity store by introducing new cache mechanisms for entities with credentials and for mapping client IDs to entity IDs. It also refines cache invalidation logic and adds comprehensive tests to ensure correct caching behavior and error handling.

### Approach

**Caching improvements:**

- Added a new cache (`entityWithCredentialsByIDCache`) to store entities along with their credentials, optimizing repeated lookups for credentialed entities. The `GetEntityWithCredentials` method now checks this cache before querying the underlying store and caches results after a miss. [[1]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R33-R47) [[2]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277L74-R107) [[3]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R315-R326) [[4]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R335-R338)
- Introduced a cache (`entityIDByClientIDCache`) to quickly resolve entity IDs from client IDs, including logic to extract and cache the mapping during entity creation and updates, and to consult the cache in `IdentifyEntity`. [[1]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R33-R47) [[2]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R60-R62) [[3]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R157-R182) [[4]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R256-R304)

**Cache invalidation and helpers:**

- Refined cache invalidation to ensure that updates or deletions to entities also invalidate both the new credentialed entity cache and the client ID cache, preventing stale data. [[1]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277L74-R107) [[2]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R127) [[3]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R157-R182) [[4]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R256-R304) [[5]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R335-R338)
- Added helper functions for extracting client IDs from entities and for handling cache set/delete operations, with appropriate error logging. [[1]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R256-R304) [[2]](diffhunk://#diff-d2e42cf48e4cc60e4b4373ad34736a12bdb474cb3170da7a14112a3d5d592277R315-R326)

**Testing enhancements:**

- Expanded the test suite to cover the new caches, including hit/miss scenarios, error handling for cache set/delete failures, and edge cases (such as nil results or entities). [[1]](diffhunk://#diff-fcf3ab5fe993b9f93441602b92dcbc9403c4de1b5f9e037f64a9cb86406801f1R40-R45) [[2]](diffhunk://#diff-fcf3ab5fe993b9f93441602b92dcbc9403c4de1b5f9e037f64a9cb86406801f1R55-R73) [[3]](diffhunk://#diff-fcf3ab5fe993b9f93441602b92dcbc9403c4de1b5f9e037f64a9cb86406801f1R168-R218) [[4]](diffhunk://#diff-fcf3ab5fe993b9f93441602b92dcbc9403c4de1b5f9e037f64a9cb86406801f1R261-R370)
- Updated tests to check that client ID mappings are cached or not cached appropriately after entity creation or store errors. [[1]](diffhunk://#diff-fcf3ab5fe993b9f93441602b92dcbc9403c4de1b5f9e037f64a9cb86406801f1R416-R420) [[2]](diffhunk://#diff-fcf3ab5fe993b9f93441602b92dcbc9403c4de1b5f9e037f64a9cb86406801f1R434-R437)

### Related Issues
- https://github.com/thunder-id/thunderid/issues/2806

### Related PRs
- N/A

### Checklist
- [x] Followed the contribution guidelines.
- [x] Manual test round performed and verified.
- [ ] Documentation provided. (Add links if there are any)
    - [ ] Ran Vale and fixed all errors and warnings
- [x] Tests provided. (Add links if there are any)
    - [x] Unit Tests
    - [ ] Integration Tests
- [ ] Breaking changes. (Fill if applicable)
    - [ ] Breaking changes section filled.
    - [ ] `breaking change` label added.

### Security checks
- [x] Followed secure coding standards in [WSO2 Secure Coding Guidelines](https://security.docs.wso2.com/en/latest/security-guidelines/secure-engineering-guidelines/secure-coding-guidlines/introduction/)
- [x] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.

## Summary by CodeRabbit

* **Refactor**
  * Added credential-aware per-entity caching and a client-ID→entity-ID mapping to reduce redundant lookups and improve lookup performance.
* **Bug Fixes**
  * Extended cache invalidation to keep client-ID and credential entries consistent across create, update, attribute changes and delete operations.
* **Tests**
  * Added tests for credential cache hits/misses, client-ID caching behavior, invalidation paths, and error edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/thunder-id/thunderid/pull/2807?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by senthalan — DISMISSED** [review]

**Review by KaveeshaPiumini — DISMISSED** [review]

File: backend/internal/entity/cache_backed_store.go [hunk 1]
```diff
@@ -71,7 +74,18 @@ func (s *cacheBackedEntityStore) GetEntity(ctx context.Context, id string) (Enti
...
+	}
+
+	result, err := s.store.GetEntityWithCredentials(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	s.cacheEntityWithCredentialsByID(ctx, result)
+	return result, nil
 }
```
[the remark this answers is not in the document]
Comment (ThumulaPerera) [reply backend/internal/entity/cache_backed_store.go:103]: no caller mutates the returned data

File: backend/internal/entity/cache_backed_store.go [hunk 2]
```diff
@@ -29,17 +29,23 @@ import (
 // cacheBackedEntityStore wraps an entityStoreInterface with in-memory caching
 // for individual entity lookups by ID and identifier filter resolution.
 type cacheBackedEntityStore struct {
-	entityByIDCache cache.CacheInterface[*Entity]
-	store           entityStoreInterface
-	logger          *log.Logger
+	entityByIDCache                cache.CacheInterface[*Entity]
+	entityWithCredentialsByIDCache cache.CacheInterface[*entityWithCredentials]
+	entityIDByClientIDCache        cache.CacheInterface[*string]
```
Comment (senthalan) [inline backend/internal/entity/cache_backed_store.go]: Can't we handle this generically ? without keeping the clientID name in here? can't we make it set up supported cachedKeys and we can have it as clientID for now

Comment (ThumulaPerera) [reply backend/internal/entity/cache_backed_store.go]: Made this generic for the case for single filter. Did not add support for multiple filter case as that will add an unnecessary complexity without a usecase as of now

File: backend/internal/entity/cache_backed_store.go [hunk 3]
```diff
@@ -125,6 +157,10 @@ func (s *cacheBackedEntityStore) UpdateSystemCredentials(ctx context.Context,
 }
 
 func (s *cacheBackedEntityStore) DeleteEntity(ctx context.Context, id string) error {
+	// Invalidate identifier cache before the store delete so the store fallback
+	// can still fetch the entity if the by-ID cache is cold.
```
Comment (senthalan) [inline backend/internal/entity/cache_backed_store.go:150]: Don't we need to do the same for the update flows ? invalidateIdentifierCache before updating the store. What if the identifier is updated? we only support client id now, but in future, if we cache other attributes it's gonna be a problem

Comment (ThumulaPerera) [reply backend/internal/entity/cache_backed_store.go:150]: Updated

**Review by senthalan — APPROVED** [review]
